smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/XmlFlattenedTraitValidator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.model.validation.validators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.XmlFlattenedTrait;
+import software.amazon.smithy.model.traits.XmlNameTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates that xmlFlattened members aren't unintentionally ignoring the
+ * xmlName of their targets.
+ */
+public final class XmlFlattenedTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (MemberShape member : model.getMemberShapesWithTrait(XmlFlattenedTrait.class)) {
+            // Don't emit the event if they're being explicit about the xmlName on this member
+            if (member.hasTrait(XmlNameTrait.class)) {
+                continue;
+            }
+
+            model.getShape(member.getTarget())
+                    .flatMap(Shape::asListShape)
+                    .flatMap(listShape -> validateMemberTargetingList(member, listShape))
+                    .ifPresent(events::add);
```

Comment:
IMO reading this as a stream is harder than just reading `if`s since we know the member target should exist.

```
Shape target = model.expectShape(member.getTarget());
if (target.isListShape()) {
    validateMemberTargetListing(member, targetList).ifPresent(events::add);
}
```

Reply:
True I forgot we run that type of validation first, so we can expect the member target to exist.